File: app/src/test/java/blue/koenig/kingsfinances/DebtsCalculatorTests.java
package blue.koenig.kingsfinances;

import com.koenig.commonModel.finance.CostDistribution;
import com.koenig.commonModel.finance.Expenses;

import junit.framework.Assert;

import org.joda.time.DateTime;
import org.joda.time.Period;
import org.junit.Test;

import java.util.ArrayList;
import java.util.List;

import blue.koenig.kingsfinances.model.calculation.DebtsCalculator;
import blue.koenig.kingsfinances.model.calculation.StatisticEntry;

import static blue.koenig.kingsfinances.TestHelper.assertDebtsList;
import static blue.koenig.kingsfinances.TestHelper.getCalculatorService;
import static blue.koenig.kingsfinances.TestHelper.getDay;
import static blue.koenig.kingsfinances.TestHelper.makeCostDistribution;
import static blue.koenig.kingsfinances.TestHelper.makeDebts;
import static blue.koenig.kingsfinances.TestHelper.makeDebtsList;
import static blue.koenig.kingsfinances.TestHelper.milena;
import static blue.koenig.kingsfinances.TestHelper.thomas;

/**
 * Example local unit test, which will execute on the development machine (host).
 *
 * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
 */
public class DebtsCalculatorTests {

    public Expenses makeExpenses(CostDistribution costDistribution, DateTime dateTime) {
        return new Expenses("", "", "", costDistribution.sumReal(), costDistribution, dateTime, "");
    }


    @Test
    public void debtsCalculation1() throws Exception {
        List<Expenses> expensesList = new ArrayList<>(1);
        expensesList.add(makeExpenses(makeCostDistribution(-10, 0, 0, -10), getDay(2017, 1, 1)));
        List<StatisticEntry> debts = DebtsCalculator.recalculateAll(expensesList, Period.months(1));
        Assert.assertEquals(2, debts.size());

        StatisticEntry debt = debts.get(0);
        Assert.assertEquals(getDay(2016, 12, 1), debt.getDate());
        Assert.assertEquals(0, debt.getEntryFor(thomas));
        Assert.assertEquals(0, debt.getEntryFor(milena));

        debt = debts.get(1);
        Assert.assertEquals(getDay(2017, 1, 1), debt.getDate());
        Assert.assertEquals(-10, debt.getEntryFor(thomas));
        Assert.assertEquals(10, debt.getEntryFor(milena));
    }

    @Test
    public void debtsCalculation2() throws Exception {
        List<Expenses> expensesList = new ArrayList<>(2);
        expensesList.add(makeExpenses(makeCostDistribution(-10, 0, 0, -10), getDay(2017, 1, 1)));
        expensesList.add(makeExpenses(makeCostDistribution(-5, -10, -5, 0), getDay(2017, 1, 31)));
        List<StatisticEntry> debts = DebtsCalculator.recalculateAll(expensesList, Period.months(1));
        Assert.assertEquals(3, debts.size());

        StatisticEntry debt = debts.get(0);
        Assert.assertEquals(getDay(2016, 12, 1), debt.getDate());
        Assert.assertEquals(0, debt.getEntryFor(thomas));
        Assert.assertEquals(0, debt.getEntryFor(milena));

        debt = debts.get(1);
        Assert.assertEquals(getDay(2017, 1, 1), debt.getDate());
        Assert.assertEquals(-10, debt.getEntryFor(thomas));
        Assert.assertEquals(10, debt.getEntryFor(milena));

        debt = debts.get(2);
        Assert.assertEquals(getDay(2017, 2, 1), debt.getDate());
        Assert.assertEquals(-5, debt.getEntryFor(thomas));
        Assert.assertEquals(5, debt.getEntryFor(milena));
    }

    @Test
    public void debtsCalculation3() throws Exception {
        List<Expenses> expensesList = new ArrayList<>(3);
        expensesList.add(makeExpenses(makeCostDistribution(-10, 0, 0, -10), getDay(2017, 1, 1)));
        expensesList.add(makeExpenses(makeCostDistribution(-10, 0, 0, -10), getDay(2017, 1, 1)));
        expensesList.add(makeExpenses(makeCostDistribution(-5, -10, -5, 0), getDay(2017, 1, 31)));
        expensesList.add(makeExpenses(makeCostDistribution(0, 300, 300, 0), getDay(2017, 3, 2)));
        List<StatisticEntry> debts = DebtsCalculator.recalculateAll(expensesList, Period.months(1));
        Assert.assertEquals(5, debts.size());

        StatisticEntry debt = debts.get(0);
        Assert.assertEquals(getDay(2016, 12, 1), debt.getDate());
        Assert.assertEquals(0, debt.getEntryFor(thomas));
        Assert.assertEquals(0, debt.getEntryFor(milena));

        debt = debts.get(1);
        Assert.assertEquals(getDay(2017, 1, 1), debt.getDate());
        Assert.assertEquals(-20, debt.getEntryFor(thomas));
        Assert.assertEquals(20, debt.getEntryFor(milena));

        debt = debts.get(2);
        Assert.assertEquals(getDay(2017, 2, 1), debt.getDate());
        Assert.assertEquals(-15, debt.getEntryFor(thomas));
        Assert.assertEquals(15, debt.getEntryFor(milena));

        debt = debts.get(3);
        Assert.assertEquals(getDay(2017, 3, 1), debt.getDate());
        Assert.assertEquals(-15, debt.getEntryFor(thomas));
        Assert.assertEquals(15, debt.getEntryFor(milena));

        debt = debts.get(4);
        Assert.assertEquals(getDay(2017, 4, 1), debt.getDate());
        Assert.assertEquals(-315, debt.getEntryFor(thomas));
        Assert.assertEquals(315, debt.getEntryFor(milena));
    }


    @Test
    public void emptyDebts() {
        List<StatisticEntry> statisticEntryList = DebtsCalculator.updateStatistics(makeDebts(-10, 17, 1, 2), Period.months(1), new ArrayList<>());
        Assert.assertEquals(2, statisticEntryList.size());

        assertDebtsList(0, getDay(17, 1, 1), 0, statisticEntryList);
        assertDebtsList(1, getDay(17, 2, 1), -10, statisticEntryList);
    }

    @Test
    public void addOneBefore() {
        List<StatisticEntry> statisticEntryList = DebtsCalculator.updateStatistics(makeDebts(-10, 17, 1, 2), Period.months(1), makeDebtsList(getDay(17, 2, 1), new int[]{0, 20}));
        Assert.assertEquals(3, statisticEntryList.size());

        assertDebtsList(0, getDay(17, 1, 1), 0, statisticEntryList);
        assertDebtsList(1, getDay(17, 2, 1), -10, statisticEntryList);
        assertDebtsList(2, getDay(17, 3, 1), 10, statisticEntryList);
    }

    @Test
    public void addBefore() {
        List<StatisticEntry> statisticEntryList = DebtsCalculator.updateStatistics(makeDebts(-10, 17, 1, 2), Period.months(1), makeDebtsList(getDay(17, 1, 1), new int[]{0, 20}));
        Assert.assertEquals(2, statisticEntryList.size());

        assertDebtsList(0, getDay(17, 1, 1), 0, statisticEntryList);
        assertDebtsList(1, getDay(17, 2, 1), 10, statisticEntryList);
    }


    @Test
    public void addWayBefore() {
        List<StatisticEntry> statisticEntryList = DebtsCalculator.updateStatistics(makeDebts(-10, 16, 6, 2), Period.months(1), makeDebtsList(getDay(17, 1, 1), new int[]{0, 20}));
        Assert.assertEquals(9, statisticEntryList.size());

        assertDebtsList(0, getDay(16, 6, 1), 0, statisticEntryList);
        assertDebtsList(1, getDay(16, 7, 1), -10, statisticEntryList);
        assertDebtsList(2, getDay(16, 8, 1), -10, statisticEntryList);
        assertDebtsList(3, getDay(16, 9, 1), -10, statisticEntryList);
        assertDebtsList(4, getDay(16, 10, 1), -10, statisticEntryList);
        assertDebtsList(5, getDay(16, 11, 1), -10, statisticEntryList);
        assertDebtsList(6, getDay(16, 12, 1), -10, statisticEntryList);
        assertDebtsList(7, getDay(17, 1, 1), -10, statisticEntryList);
        assertDebtsList(8, getDay(17, 2, 1), 10, statisticEntryList);
    }

    @Test
    public void addMiddle() {
        List<StatisticEntry> statisticEntryList = DebtsCalculator.updateStatistics(makeDebts(-10, 16, 12, 2), Period.months(1),
                makeDebtsList(getDay(16, 6, 1), new int[]{0, 20, 30, 40, 40, 40, 50, -20, 0}));
        Assert.assertEquals(9, statisticEntryList.size());

        assertDebtsList(0, getDay(16, 6, 1), 0, statisticEntryList);
        assertDebtsList(1, getDay(16, 7, 1), 20, statisticEntryList);
        assertDebtsList(2, getDay(16, 8, 1), 30, statisticEntryList);
        assertDebtsList(3, getDay(16, 9, 1), 40, statisticEntryList);
        assertDebtsList(4, getDay(16, 10, 1), 40, statisticEntryList);
        assertDebtsList(5, getDay(16, 11, 1), 40, statisticEntryList);
        assertDebtsList(6, getDay(16, 12, 1), 50, statisticEntryList);
        assertDebtsList(7, getDay(17, 1, 1), -30, statisticEntryList);
        assertDebtsList(8, getDay(17, 2, 1), -10, statisticEntryList);
    }

    @Test
    public void addOneEnd() {
        List<StatisticEntry> statisticEntryList = DebtsCalculator.updateStatistics(makeDebts(-10, 17, 2, 2), Period.months(1),
                makeDebtsList(getDay(17, 1, 1), new int[]{0, 20}));
        Assert.assertEquals(3, statisticEntryList.size());

        assertDebtsList(0, getDay(17, 1, 1), 0, statisticEntryList);
        assertDebtsList(1, getDay(17, 2, 1), 20, statisticEntryList);
        assertDebtsList(2, getDay(17, 3, 1), 10, statisticEntryList);
    }

    @Test
    public void addWayEnd() {
        List<StatisticEntry> statisticEntryList = DebtsCalculator.updateStatistics(makeDebts(-10, 17, 4, 2), Period.months(1),
                makeDebtsList(getDay(17, 1, 1), new int[]{0, 20}));
        Assert.assertEquals(5, statisticEntryList.size());

        assertDebtsList(0, getDay(17, 1, 1), 0, statisticEntryList);
        assertDebtsList(1, getDay(17, 2, 1), 20, statisticEntryList);
        assertDebtsList(2, getDay(17, 3, 1), 20, statisticEntryList);
        assertDebtsList(3, getDay(17, 4, 1), 20, statisticEntryList);
        assertDebtsList(4, getDay(17, 5, 1), 10, statisticEntryList);
    }


    @Test
    public void editFirst() {
        TestExpensesSubject expensesItemSubject = new TestExpensesSubject();
        DebtsCalculator debtsCalculator = new DebtsCalculator(Period.months(1), expensesItemSubject,
                getCalculatorService(makeDebtsList(getDay(17, 1, 1), new int[]{0, 0, 0, -10})));
        expensesItemSubject.updateDebts(makeDebts(getDay(17, 1, 2), 0), makeDebts(getDay(17, 1, 2), 1));
        List<StatisticEntry> statisticEntryList = debtsCalculator.getEntrys();

        Assert.assertEquals(4, statisticEntryList.size());

        assertDebtsList(0, getDay(17, 1, 1), 0, statisticEntryList);
        assertDebtsList(1, getDay(17, 2, 1), 1, statisticEntryList);
        assertDebtsList(2, getDay(17, 3, 1), 1, statisticEntryList);
        assertDebtsList(3, getDay(17, 4, 1), -9, statisticEntryList);
    }

    @Test
    public void editDateFromFirst() {
        TestExpensesSubject expensesItemSubject = new TestExpensesSubject();
        DebtsCalculator debtsCalculator = new DebtsCalculator(Period.months(1), expensesItemSubject,
                getCalculatorService(makeDebtsList(getDay(17, 1, 1), new int[]{0, 10, 0, -10})));
        expensesItemSubject.updateDebts(makeDebts(getDay(17, 1, 2), 10), makeDebts(getDay(17, 4, 2), 10));
        List<StatisticEntry> statisticEntryList = debtsCalculator.getEntrys();

        Assert.assertEquals(5, statisticEntryList.size());

        assertDebtsList(0, getDay(17, 1, 1), 0, statisticEntryList);
        assertDebtsList(1, getDay(17, 2, 1), 0, statisticEntryList);
        assertDebtsList(2, getDay(17, 3, 1), -10, statisticEntryList);
        assertDebtsList(3, getDay(17, 4, 1), -20, statisticEntryList);
        assertDebtsList(4, getDay(17, 5, 1), -10, statisticEntryList);
    }


}

File: app/src/test/java/blue/koenig/kingsfinances/DebtsCalculatorTests.kt
package blue.koenig.kingsfinances

import blue.koenig.kingsfinances.model.calculation.DebtsCalculator
import com.koenig.commonModel.Repository.MonthStatisticsRepository
import com.koenig.commonModel.finance.CostDistribution
import com.koenig.commonModel.finance.Expenses
import io.reactivex.Observable
import org.joda.time.LocalDate
import org.joda.time.YearMonth
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.mockito.Mock
import org.mockito.MockitoAnnotations

/**
 * Example local unit test, which will execute on the development machine (host).
 *
 * @see [Testing documentation](http://d.android.com/tools/testing)
 */

class DebtsCalculatorTests {

    val itemSubject: TestExpensesSubject = TestExpensesSubject()
    lateinit var calculator: DebtsCalculator
    var endDate: YearMonth = YearMonth(2017, 12)

    @Mock
    lateinit var deltaMonthStatisticsRepository: MonthStatisticsRepository
    @Mock
    lateinit var absoluteMonthStatisticsRepository: MonthStatisticsRepository

    fun makeExpenses(costDistribution: CostDistribution, dateTime: LocalDate): Expenses {
        return Expenses("", "", "", costDistribution.sumReal(), costDistribution, dateTime, "")
    }

    @Before
    fun setup() {
        MockitoAnnotations.initMocks(this)

    }

    fun initTest() {
        calculator = DebtsCalculator(itemSubject, deltaMonthStatisticsRepository, absoluteMonthStatisticsRepository, Observable.just(endDate))
    }


    @Test
    @Throws(Exception::class)
    fun debtsCalculation1() {
        endDate = YearMonth(2017, 1)
        initTest()
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(-10, 0, 0, -10), LocalDate(2017, 1, 1)))
        val debts = calculator.absoluteMap
        Assert.assertEquals(2, debts.size)

        var debt = debts[YearMonth(2016, 12)]!!
        Assert.assertEquals(YearMonth(2016, 12), debt.month)
        Assert.assertEquals(0, debt[TestHelper.thomas])
        Assert.assertEquals(0, debt[TestHelper.milena])

        debt = debts[YearMonth(2017, 1)]!!
        Assert.assertEquals(YearMonth(2017, 1), debt.month)
        Assert.assertEquals(-10, debt[TestHelper.thomas])
        Assert.assertEquals(10, debt[TestHelper.milena])
    }

    @Test
    @Throws(Exception::class)
    fun debtsCalculation2() {
        endDate = YearMonth(2017, 1)
        initTest()
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(-10, 0, 0, -10), LocalDate(2017, 1, 1)))
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(-5, -10, -5, 0), LocalDate(2017, 1, 31)))
        val debts = calculator.absoluteMap
        Assert.assertEquals(2, debts.size)

        var debt = debts[YearMonth(2016, 12)]!!
        Assert.assertEquals(YearMonth(2016, 12), debt.month)
        Assert.assertEquals(0, debt[TestHelper.thomas])
        Assert.assertEquals(0, debt[TestHelper.milena])

        debt = debts[YearMonth(2017, 1)]!!
        Assert.assertEquals(YearMonth(2017, 1), debt.month)
        Assert.assertEquals(-5, debt[TestHelper.thomas])
        Assert.assertEquals(5, debt[TestHelper.milena])


    }

    @Test
    @Throws(Exception::class)
    fun debtsCalculation3() {
        endDate = YearMonth(2017, 3)
        initTest()
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(-10, 0, 0, -10), LocalDate(2017, 1, 1)))
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(-10, 0, 0, -10), LocalDate(2017, 1, 1)))
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(-5, -10, -5, 0), LocalDate(2017, 1, 31)))
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(0, 300, 300, 0), LocalDate(2017, 3, 2)))
        val debts = calculator.absoluteMap
        Assert.assertEquals(4, debts.size)


        var debt = debts[YearMonth(2016, 12)]!!
        Assert.assertEquals(YearMonth(2016, 12), debt.month)
        Assert.assertEquals(0, debt[TestHelper.thomas])
        Assert.assertEquals(0, debt[TestHelper.milena])

        debt = debts[YearMonth(2017, 1)]!!
        Assert.assertEquals(YearMonth(2017, 1), debt.month)
        Assert.assertEquals(-15, debt[TestHelper.thomas])
        Assert.assertEquals(15, debt[TestHelper.milena])

        debt = debts[YearMonth(2017, 2)]!!
        Assert.assertEquals(YearMonth(2017, 2), debt.month)
        Assert.assertEquals(-15, debt[TestHelper.thomas])
        Assert.assertEquals(15, debt[TestHelper.milena])

        debt = debts[YearMonth(2017, 3)]!!
        Assert.assertEquals(YearMonth(2017, 3), debt.month)
        Assert.assertEquals(-315, debt[TestHelper.thomas])
        Assert.assertEquals(315, debt[TestHelper.milena])


    }


    @Test
    fun addOneBefore() {
        endDate = YearMonth(2017, 1)
        initTest()
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(-10, 0, 0, -10), LocalDate(2017, 1, 1)))
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(-10, 0, 0, -10), LocalDate(2017, 1, 1)))
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(-5, -10, -5, 0), LocalDate(2017, 1, 31)))
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(0, -25, -25, 0), LocalDate(2016, 12, 2)))
        val debts = calculator.absoluteMap
        Assert.assertEquals(3, debts.size)


        var debt = debts[YearMonth(2016, 12)]!!
        Assert.assertEquals(YearMonth(2016, 12), debt.month)
        Assert.assertEquals(25, debt[TestHelper.thomas])
        Assert.assertEquals(-25, debt[TestHelper.milena])

        debt = debts[YearMonth(2017, 1)]!!
        Assert.assertEquals(YearMonth(2017, 1), debt.month)
        Assert.assertEquals(10, debt[TestHelper.thomas])
        Assert.assertEquals(-10, debt[TestHelper.milena])



        debt = debts[YearMonth(2016, 11)]!!
        Assert.assertEquals(YearMonth(2016, 11), debt.month)
        Assert.assertEquals(0, debt[TestHelper.thomas])
        Assert.assertEquals(0, debt[TestHelper.milena])
    }

    @Test
    fun addBefore() {

        endDate = YearMonth(2017, 1)
        initTest()
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(-10, 0, 0, -10), LocalDate(2017, 1, 1)))
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(-10, 0, 0, -10), LocalDate(2017, 1, 1)))
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(-5, -10, -5, 0), LocalDate(2017, 1, 31)))
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(0, -25, -25, 0), LocalDate(2016, 11, 2)))
        val debts = calculator.absoluteMap
        Assert.assertEquals(4, debts.size)


        var debt = debts[YearMonth(2016, 12)]!!
        Assert.assertEquals(YearMonth(2016, 12), debt.month)
        Assert.assertEquals(25, debt[TestHelper.thomas])
        Assert.assertEquals(-25, debt[TestHelper.milena])

        debt = debts[YearMonth(2017, 1)]!!
        Assert.assertEquals(YearMonth(2017, 1), debt.month)
        Assert.assertEquals(10, debt[TestHelper.thomas])
        Assert.assertEquals(-10, debt[TestHelper.milena])



        debt = debts[YearMonth(2016, 11)]!!
        Assert.assertEquals(YearMonth(2016, 11), debt.month)
        Assert.assertEquals(25, debt[TestHelper.thomas])
        Assert.assertEquals(-25, debt[TestHelper.milena])

        debt = debts[YearMonth(2016, 10)]!!
        Assert.assertEquals(YearMonth(2016, 10), debt.month)
        Assert.assertEquals(0, debt[TestHelper.thomas])
        Assert.assertEquals(0, debt[TestHelper.milena])


    }


    @Test
    fun addWayBefore() {

        endDate = YearMonth(2017, 1)
        initTest()
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(-10, 0, 0, -10), LocalDate(2017, 1, 1)))
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(-10, 0, 0, -10), LocalDate(2017, 1, 1)))
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(-5, -10, -5, 0), LocalDate(2017, 1, 31)))
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(0, -25, -25, 0), LocalDate(2016, 9, 2)))
        val debts = calculator.absoluteMap
        Assert.assertEquals(6, debts.size)


        var debt = debts[YearMonth(2016, 12)]!!
        Assert.assertEquals(YearMonth(2016, 12), debt.month)
        Assert.assertEquals(25, debt[TestHelper.thomas])
        Assert.assertEquals(-25, debt[TestHelper.milena])

        debt = debts[YearMonth(2017, 1)]!!
        Assert.assertEquals(YearMonth(2017, 1), debt.month)
        Assert.assertEquals(10, debt[TestHelper.thomas])
        Assert.assertEquals(-10, debt[TestHelper.milena])



        debt = debts[YearMonth(2016, 11)]!!
        Assert.assertEquals(YearMonth(2016, 11), debt.month)
        Assert.assertEquals(25, debt[TestHelper.thomas])
        Assert.assertEquals(-25, debt[TestHelper.milena])

        debt = debts[YearMonth(2016, 10)]!!
        Assert.assertEquals(YearMonth(2016, 10), debt.month)
        Assert.assertEquals(25, debt[TestHelper.thomas])
        Assert.assertEquals(-25, debt[TestHelper.milena])

        debt = debts[YearMonth(2016, 9)]!!
        Assert.assertEquals(YearMonth(2016, 9), debt.month)
        Assert.assertEquals(25, debt[TestHelper.thomas])
        Assert.assertEquals(-25, debt[TestHelper.milena])

        debt = debts[YearMonth(2016, 8)]!!
        Assert.assertEquals(YearMonth(2016, 8), debt.month)
        Assert.assertEquals(0, debt[TestHelper.thomas])
        Assert.assertEquals(0, debt[TestHelper.milena])

    }

    @Test
    fun addMiddle() {
        endDate = YearMonth(2017, 3)
        initTest()
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(-10, 0, 0, -10), LocalDate(2017, 1, 1)))
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(-10, 0, 0, -10), LocalDate(2017, 3, 1)))
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(-5, -10, -5, 0), LocalDate(2017, 2, 28)))
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(0, -25, -25, 0), LocalDate(2016, 12, 2)))
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(0, -25, -25, 0), LocalDate(2017, 1, 2)))
        val debts = calculator.absoluteMap
        Assert.assertEquals(5, debts.size)

        var debt = debts[YearMonth(2016, 11)]!!
        Assert.assertEquals(YearMonth(2016, 11), debt.month)
        Assert.assertEquals(0, debt[TestHelper.thomas])
        Assert.assertEquals(0, debt[TestHelper.milena])

        debt = debts[YearMonth(2016, 12)]!!
        Assert.assertEquals(YearMonth(2016, 12), debt.month)
        Assert.assertEquals(25, debt[TestHelper.thomas])
        Assert.assertEquals(-25, debt[TestHelper.milena])

        debt = debts[YearMonth(2017, 1)]!!
        Assert.assertEquals(YearMonth(2017, 1), debt.month)
        Assert.assertEquals(40, debt[TestHelper.thomas])
        Assert.assertEquals(-40, debt[TestHelper.milena])

        debt = debts[YearMonth(2017, 2)]!!
        Assert.assertEquals(YearMonth(2017, 2), debt.month)
        Assert.assertEquals(45, debt[TestHelper.thomas])
        Assert.assertEquals(-45, debt[TestHelper.milena])

        debt = debts[YearMonth(2017, 3)]!!
        Assert.assertEquals(YearMonth(2017, 3), debt.month)
        Assert.assertEquals(35, debt[TestHelper.thomas])
        Assert.assertEquals(-35, debt[TestHelper.milena])
    }

    @Test
    fun addOneEnd() {
        endDate = YearMonth(2017, 2)
        initTest()
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(-10, 0, 0, -10), LocalDate(2017, 1, 1)))
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(-5, -10, -5, 0), LocalDate(2017, 1, 31)))
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(-5, -10, -5, 0), LocalDate(2017, 2, 1)))
        val debts = calculator.absoluteMap
        Assert.assertEquals(3, debts.size)

        var debt = debts[YearMonth(2016, 12)]!!
        Assert.assertEquals(YearMonth(2016, 12), debt.month)
        Assert.assertEquals(0, debt[TestHelper.thomas])
        Assert.assertEquals(0, debt[TestHelper.milena])

        debt = debts[YearMonth(2017, 1)]!!
        Assert.assertEquals(YearMonth(2017, 1), debt.month)
        Assert.assertEquals(-5, debt[TestHelper.thomas])
        Assert.assertEquals(5, debt[TestHelper.milena])

        debt = debts[YearMonth(2017, 2)]!!
        Assert.assertEquals(YearMonth(2017, 2), debt.month)
        Assert.assertEquals(0, debt[TestHelper.thomas])
        Assert.assertEquals(0, debt[TestHelper.milena])

    }

    @Test
    fun addWayEnd() {
        endDate = YearMonth(2017, 4)
        initTest()
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(-10, 0, 0, -10), LocalDate(2017, 1, 1)))
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(-5, -10, -5, 0), LocalDate(2017, 1, 31)))
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(-5, -10, -5, 0), LocalDate(2017, 4, 1)))
        val debts = calculator.absoluteMap
        Assert.assertEquals(5, debts.size)

        var debt = debts[YearMonth(2016, 12)]!!
        Assert.assertEquals(YearMonth(2016, 12), debt.month)
        Assert.assertEquals(0, debt[TestHelper.thomas])
        Assert.assertEquals(0, debt[TestHelper.milena])

        debt = debts[YearMonth(2017, 1)]!!
        Assert.assertEquals(YearMonth(2017, 1), debt.month)
        Assert.assertEquals(-5, debt[TestHelper.thomas])
        Assert.assertEquals(5, debt[TestHelper.milena])

        debt = debts[YearMonth(2017, 2)]!!
        Assert.assertEquals(YearMonth(2017, 2), debt.month)
        Assert.assertEquals(-5, debt[TestHelper.thomas])
        Assert.assertEquals(5, debt[TestHelper.milena])

        debt = debts[YearMonth(2017, 3)]!!
        Assert.assertEquals(YearMonth(2017, 3), debt.month)
        Assert.assertEquals(-5, debt[TestHelper.thomas])
        Assert.assertEquals(5, debt[TestHelper.milena])

        debt = debts[YearMonth(2017, 4)]!!
        Assert.assertEquals(YearMonth(2017, 4), debt.month)
        Assert.assertEquals(0, debt[TestHelper.thomas])
        Assert.assertEquals(0, debt[TestHelper.milena])

    }


    @Test
    fun editFirst() {
        endDate = YearMonth(2017, 2)
        initTest()
        val first = makeExpenses(TestHelper.makeCostDistribution(-10, 0, 0, -10), LocalDate(2017, 1, 1))
        itemSubject.add(first)
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(-5, -10, -5, 0), LocalDate(2017, 1, 31)))
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(-5, -10, -5, 0), LocalDate(2017, 2, 1)))
        itemSubject.update(first, makeExpenses(TestHelper.makeCostDistribution(-10, -10, 0, 0), LocalDate(2017, 1, 1)))
        val debts = calculator.absoluteMap
        Assert.assertEquals(3, debts.size)

        var debt = debts[YearMonth(2016, 12)]!!
        Assert.assertEquals(YearMonth(2016, 12), debt.month)
        Assert.assertEquals(0, debt[TestHelper.thomas])
        Assert.assertEquals(0, debt[TestHelper.milena])

        debt = debts[YearMonth(2017, 1)]!!
        Assert.assertEquals(YearMonth(2017, 1), debt.month)
        Assert.assertEquals(5, debt[TestHelper.thomas])
        Assert.assertEquals(-5, debt[TestHelper.milena])

        debt = debts[YearMonth(2017, 2)]!!
        Assert.assertEquals(YearMonth(2017, 2), debt.month)
        Assert.assertEquals(10, debt[TestHelper.thomas])
        Assert.assertEquals(-10, debt[TestHelper.milena])
    }

    @Test
    fun editDateFromFirst() {
        endDate = YearMonth(2017, 2)
        initTest()
        val first = makeExpenses(TestHelper.makeCostDistribution(-10, 0, 0, -10), LocalDate(2017, 1, 1))
        itemSubject.add(first)
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(-5, -10, -5, 0), LocalDate(2017, 1, 31)))
        itemSubject.add(makeExpenses(TestHelper.makeCostDistribution(-5, -10, -5, 0), LocalDate(2017, 2, 1)))
        itemSubject.update(first, makeExpenses(TestHelper.makeCostDistribution(-5, -10, -5, 0), LocalDate(2016, 12, 1)))
        val debts = calculator.absoluteMap
        Assert.assertEquals(4, debts.size)

        var debt = debts[YearMonth(2016, 11)]!!
        Assert.assertEquals(YearMonth(2016, 11), debt.month)
        Assert.assertEquals(0, debt[TestHelper.thomas])
        Assert.assertEquals(0, debt[TestHelper.milena])

        debt = debts[YearMonth(2016, 12)]!!
        Assert.assertEquals(YearMonth(2016, 12), debt.month)
        Assert.assertEquals(5, debt[TestHelper.thomas])
        Assert.assertEquals(-5, debt[TestHelper.milena])

        debt = debts[YearMonth(2017, 1)]!!
        Assert.assertEquals(YearMonth(2017, 1), debt.month)
        Assert.assertEquals(10, debt[TestHelper.thomas])
        Assert.assertEquals(-10, debt[TestHelper.milena])

        debt = debts[YearMonth(2017, 2)]!!
        Assert.assertEquals(YearMonth(2017, 2), debt.month)
        Assert.assertEquals(15, debt[TestHelper.thomas])
        Assert.assertEquals(-15, debt[TestHelper.milena])
    }

}